compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,838 @@
+//! assignment that is captured by the analysis, all mutable access to the underlying place (which
+//! is not observable by the analysis) must be invalidated. This is (hopefully) covered by Stacked
+//! Borrows.
+//!
+//! To be continued...
+
+use std::fmt::{Debug, Formatter};
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Ty, TyCtxt};
+use rustc_target::abi::VariantIdx;
+
+use crate::{
+    fmt::DebugWithContext, lattice::FlatSet, Analysis, AnalysisDomain, CallReturnPlaces,
+    JoinSemiLattice, SwitchIntEdgeEffects,
+};
+
+pub trait ValueAnalysis<'tcx> {
+    /// For each place of interest, the analysis tracks a value of the given type.
+    type Value: Clone + JoinSemiLattice + HasBottom + HasTop;
+
+    const NAME: &'static str;
+
+    fn map(&self) -> &Map;
+
+    fn handle_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        self.super_statement(statement, state)
+    }
+
+    fn super_statement(&self, statement: &Statement<'tcx>, state: &mut State<Self::Value>) {
+        match &statement.kind {
+            StatementKind::Assign(box (place, rvalue)) => {
+                self.handle_assign(*place, rvalue, state);
+            }
+            StatementKind::SetDiscriminant { .. } => {
+                // Could treat this as writing a constant to a pseudo-place.
+            }
+            StatementKind::CopyNonOverlapping(..) => {
+                // FIXME: What to do here?
+            }
+            StatementKind::StorageLive(..)
+            | StatementKind::StorageDead(..)
+            | StatementKind::Deinit(_) => {
+                // Could perhaps use these.
+            }
+            StatementKind::Nop
+            | StatementKind::Retag(..)
+            | StatementKind::FakeRead(..)
+            | StatementKind::Coverage(..)
+            | StatementKind::AscribeUserType(..) => (),
+        }
+    }
+
+    fn handle_assign(
+        &self,
+        target: Place<'tcx>,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) {
+        self.super_assign(target, rvalue, state)
+    }
+
+    fn super_assign(
+        &self,
+        target: Place<'tcx>,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) {
+        let result = self.handle_rvalue(rvalue, state);
+        state.assign(target.as_ref(), result, self.map());
+    }
+
+    fn handle_rvalue(
+        &self,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) -> ValueOrPlaceOrRef<Self::Value> {
+        self.super_rvalue(rvalue, state)
+    }
+
+    fn super_rvalue(
+        &self,
+        rvalue: &Rvalue<'tcx>,
+        state: &mut State<Self::Value>,
+    ) -> ValueOrPlaceOrRef<Self::Value> {
+        match rvalue {
+            Rvalue::Use(operand) => self.handle_operand(operand, state).into(),
+            Rvalue::Ref(_, BorrowKind::Shared, place) => self
```

Comment:
Here is where you can put a check that `place.ty()` is `Freeze`. You'll want to use `Ty::freeze` for this. That will require a `ParamEnv` computed via `tcx.param_env_reveal_all_normalized(function_def_id)` and a `LocalDecls`, which you'll need to get access to somehow.

Reply:
I'm not sure whether this suffices. Let's assume that an implementation tracks a non-top value contained in an `UnsafeCell`. Even if we flood the place and return `Unknown` here, the resulting reference can still be used to change the underlying value. And if a value was tracked inside that `UnsafeCell`, the analysis can probably enter another non-top value down the road. This would then violate the invariant that the value must be top if mutable access can occur without observation.

Perhaps we should prefer denying the registration of places that are `!Freeze`? Also, failing early might be better.

Reply:
> the resulting reference can still be used to change the underlying value

Yes. This is true in the `&mut` case as well.

> And if a value was tracked inside that UnsafeCell, the analysis can probably enter another non-top value down the road.

Assuming the type of the local is `UnsafeCell<i32>`, this would look like `_1.inner = value`. The pointer used for the access here is the one that is implicitly associated with the place itself, which always has `Unique` permissions in SB. This means any pointers above it are invalidated, including the previously created one.

Reply:
> The specific thing that matters for our purposes is the assumption that an assignment to a place through a local (so like _5.a.b.c) invalidates all current pointers to that place. What I'm saying above is that under SB, this is true even considering UnsafeCell, and so we have nothing to worry about.

Cc https://github.com/rust-lang/unsafe-code-guidelines/issues/363
IMO we have a lot to worry about since we shouldn't be relying on Stacked Borrows details until we are sure that is the model we want.

For now, if a pointer or reference of any kind (including a shared ref) is ever created to a place, I think this analysis should consider that place permanently tainted -- as would be the case if we had no aliasing rules whatsoever.

Reply:
> Here is where you can put a check that `place.ty()` is `Freeze`. You'll want to use `Ty::freeze` for this. That will require a `ParamEnv` computed via `tcx.param_env_reveal_all_normalized(function_def_id)` and a `LocalDecls`, which you'll need to get access to somehow.

I think I solved this issue. I did not find a nice way to put the `Freeze` check where you suggested it. Instead, when determining which places to track, we only consider dereferences of a place if the target type is `Freeze`. This should effectively do the same as your suggestion, but additionally prevents adding places to `Map` which can never get values assigned anyway.